In [142]:
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

import tensorflow
from tensorflow import keras
from keras.layers import Dense, Dropout, BatchNormalization
from keras.models import Sequential
from keras.optimizers import SGD

from sklearn.metrics import r2_score

In [143]:
df=pd.read_csv("/content/boston.csv")
df.head()

CRIM    ZN  INDUS  CHAS     NX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296.0   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242.0   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242.0   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222.0   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222.0   

   PTRATIO       B  LSTAT  MEDV  
0     15.3  396.90   4.98  24.0  
1     17.8  396.90   9.14  21.6  
2     17.8  392.83   4.03  34.7  
3     18.7  394.63   2.94  33.4  
4     18.7  396.90   5.33  36.2

In [144]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     506 non-null    float64
 1   ZN       506 non-null    float64
 2   INDUS    506 non-null    float64
 3   CHAS     506 non-null    int64  
 4   NX       506 non-null    float64
 5   RM       506 non-null    float64
 6   AGE      506 non-null    float64
 7   DIS      506 non-null    float64
 8   RAD      506 non-null    int64  
 9   TAX      506 non-null    float64
 10  PTRATIO  506 non-null    float64
 11  B        506 non-null    float64
 12  LSTAT    506 non-null    float64
 13  MEDV     506 non-null    float64
dtypes: float64(12), int64(2)
memory usage: 55.5 KB


In [145]:
df.isnull().sum()

CRIM       0
ZN         0
INDUS      0
CHAS       0
NX         0
RM         0
AGE        0
DIS        0
RAD        0
TAX        0
PTRATIO    0
B          0
LSTAT      0
MEDV       0
dtype: int64

In [146]:
x=df.iloc[:,:-1]
y=df["MEDV"]

In [147]:
x.head(2)

CRIM    ZN  INDUS  CHAS     NX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296.0   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242.0   

   PTRATIO      B  LSTAT  
0     15.3  396.9   4.98  
1     17.8  396.9   9.14

In [148]:
y.head(2)

0    24.0
1    21.6
Name: MEDV, dtype: float64

In [149]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=1)

In [150]:
sc = StandardScaler()
x_train = sc.fit_transform(x_train)
x_test=sc.transform(x_test)

In [151]:
x_train[0]

array([-0.3892494 , -0.49559343, -0.60928978, -0.29329423, -0.89958293,
       -0.14496758, -2.15002959,  0.89445528, -0.74632984, -1.00850765,
       -0.24857777,  0.28674182, -0.96685016])

In [152]:
x_test[0]

array([-0.3993398 ,  0.92318511, -1.30081723, -0.29329423, -0.69401534,
        0.8428659 ,  0.08287899, -0.30372877, -0.28494154, -1.07366536,
       -0.01782083,  0.43472666, -0.72820873])

In [153]:
model=Sequential()
model.add(Dense(input_dim=13, units=16, activation="relu"))
model.add(Dense(1,activation='linear'))
model.summary()

Model: "sequential_16"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_32 (Dense)            (None, 16)                224       
                                                                 
 dense_33 (Dense)            (None, 1)                 17        
                                                                 
Total params: 241 (964.00 Byte)
Trainable params: 241 (964.00 Byte)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [154]:
model.compile(optimizer='sgd',loss="mean_squared_error")

In [155]:
model.fit(x_train,y_train,epochs=50)

Epoch 1/50
13/13 [==============================] - 0s 2ms/step - loss: 175.3021
Epoch 2/50
13/13 [==============================] - 0s 2ms/step - loss: 28.7572
Epoch 3/50
13/13 [==============================] - 0s 2ms/step - loss: 24.0332
Epoch 4/50
13/13 [==============================] - 0s 2ms/step - loss: 20.7576
Epoch 5/50
13/13 [==============================] - 0s 2ms/step - loss: 21.1024
Epoch 6/50
13/13 [==============================] - 0s 2ms/step - loss: 18.0106
Epoch 7/50
13/13 [==============================] - 0s 2ms/step - loss: 17.0515
Epoch 8/50
13/13 [==============================] - 0s 2ms/step - loss: 15.2974
Epoch 9/50
13/13 [==============================] - 0s 2ms/step - loss: 15.7328
Epoch 10/50
13/13 [==============================] - 0s 2ms/step - loss: 14.7297
Epoch 11/50
13/13 [==============================] - 0s 2ms/step - loss: 15.5812
Epoch 12/50
13/13 [==============================] - 0s 2ms/step - loss: 13.6816
Epoch 13/50
13/13 [=================

In [156]:
y_pred=model.predict(x_test)

4/4 [==============================] - 0s 2ms/step


In [157]:
print(y_test[:5])
print('***')
print(y_pred[:5])

307    28.2
343    23.9
47     16.6
67     22.0
362    20.8
Name: MEDV, dtype: float64
***
[[29.046461]
 [22.700516]
 [20.031387]
 [20.159557]
 [24.913979]]


In [158]:
round(r2_score(y_pred,y_test),4)

0.9158

## R2 score of 0.9158 indicates that the model explains approximately 91.58% of the variance in the dependent variable (target) based on the independent variables (features) for the test dataset.